core-patch/src/main/scala/eu/ostrzyciel/jelly/core/patch/internal/PatchDecoderImpl.scala

```diff
@@ -0,0 +1,188 @@
+package eu.ostrzyciel.jelly.core.patch.internal
+
+import eu.ostrzyciel.jelly.core.JellyExceptions.RdfProtoDeserializationError
+import eu.ostrzyciel.jelly.core.internal.*
+import eu.ostrzyciel.jelly.core.*
+import eu.ostrzyciel.jelly.core.patch.*
+import eu.ostrzyciel.jelly.core.patch.handler.*
+import eu.ostrzyciel.jelly.core.proto.v1.*
+import eu.ostrzyciel.jelly.core.proto.v1.patch.*
+
+import scala.annotation.{experimental, switch}
+import scala.compiletime.uninitialized
+import scala.reflect.ClassTag
+
+@experimental
+sealed abstract class PatchDecoderImpl[TNode, TDatatype : ClassTag](
+  protected override val converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
+  handler: PatchHandler[TNode],
+  supportedOptions: RdfPatchOptions,
+) extends PatchDecoder, ProtoDecoderBase[TNode, TDatatype, Any, Any]:
+
+  private var streamOpt: Option[RdfPatchOptions] = None
+  private var isFrameStreamType: Boolean = false
+  private var isPunctuatedStreamType: Boolean = false
+
+  protected final override def getNameTableSize: Int =
+    streamOpt.map(_.maxNameTableSize) getOrElse JellyOptions.smallNameTableSize
+  protected final override def getPrefixTableSize: Int =
+    streamOpt.map(_.maxPrefixTableSize) getOrElse JellyOptions.smallPrefixTableSize
+  protected final override def getDatatypeTableSize: Int =
+    streamOpt.map(_.maxDatatypeTableSize) getOrElse 20
+
+  override def getPatchOpt: Option[RdfPatchOptions] = streamOpt
+
+  private final def setPatchOpt(opt: RdfPatchOptions): Unit =
+    if streamOpt.isEmpty then
+      streamOpt = Some(opt)
+      isFrameStreamType = opt.streamType.isFrame
+      isPunctuatedStreamType = opt.streamType.isPunctuated
+
+  override final def ingestFrame(frame: RdfPatchFrame): Unit =
+    for row <- frame.rows do
+      ingestRow(row)
+    if isFrameStreamType then
```

Comment:
yeah, so in this case this checks out and handler.punctuation() is called, but in the trait I would phrase it differently, more in terms of shoulds / do sth similar to the run/doRun pattern to ensure it is always called if needed.

Reply:
I like it this way. Making a pattern like you described would add overhead, we can't have that here. Keep in mind that this code is very different from jelly-cli, having very high cyclomatic test coverage, and this behavior is covered there.